### ISSUE TITLE
Improve versioning for minor releases

### ISSUE THREAD

**Issue by ma2ciek** [ckeditor/ckeditor5#7323]
## 📝 Provide a description of the improvement

After having the first minor release with the [new versioning policy](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/support/versioning-policy.html) I found that the current solution have a few issues:

*   The release is marked as 19.1.0, but not all packages are at this version. I imagine people changing all 19.0.0 to 19.1.0 and installation errors. This make the update a little bit tough for consumers. There're some tools like `ncu` etc. but not everyone is familiar with them and IMO this knowledge should not be necessary to update CKEditor 5.
*   It's hard to pin the 19.0.0 version without removing `^` from all CKEditor 5 dependency version ranges. And I'm afraid that this might not work in some cases as the 19.0.0 version of package A will require B@19.0.1, which might depend on the C@^19.1.0 package which was installed at version 19.0.0 already in the root, which will lead to a dup error.

After talking with @Reinmar we figure out that it can be solved in two ways:

### 1\. Release all packages during minor releases at the same version

It can be implemented similarly to how we do major releases. We should do a minor release of all packages if any package contains new feature or a minor breaking change. This way we'll solve the two above issues (2nd issue will be solvable by the `~19.1.0` range which we don't support with the current versioning - our versioning docs misses this information).

### 2\. Release all CKEditor 5 packages with the same version for all releases

IDK how often we'll be doing patch releases, but this would improve the situation there as well. There're though some important questions to answer if we go towards this direction:

*   What should we do if package A should have a bugfix after the release? In that case we can skip the version in all packages and release the new version skipping the version (I guess it's a very rare situation) For non-patch releases We'll skip the `X.Y.1` version by creating `X+1.Y.0` or `X.Y+1.0` releases
*   Should we align versioning in CF and CKE5? I think that whenever we can we should create the same releases. Since NPM packages for both projects names starts from `@ckeditor/ckeditor...` it's easier for our clients to version these packages. Other than that nothing should happen if we ship e.g. CF@20.0.0 major release together with CKE5 19.2.0 minor release as CF depend on the CKE5, not vice-versa. When CKE5 creates a major release then CF and CS should create a major release to not create duplication problem, the same probably should be implemented for minor release.

That's an important topic since minor breaking changes might break someone's code.

cc @Reinmar @czerwonkabartosz @scofalik @pomek @mlewand 

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

**Comment by fredck** [comment]
I’ve just upgraded GitHub Writer to v19.1.1 and I believe I immediately faced a problem related to the new versioning.

For example, one of the packages that bumped version is ckeditor5-utils, which is now v19.0.2. The problem is that project dependencies on this package haven’t been updated, so we have deps pointing to different versions… a few examples:

*   [root -> v19.0.2](https://github.com/ckeditor/ckeditor5/blob/893b873002c838fb10efda35515ccfaf44c3e39a/package.json#L73)
*   [core -> v19.0.1](https://github.com/ckeditor/ckeditor5/blob/893b873002c838fb10efda35515ccfaf44c3e39a/packages/ckeditor5-core/package.json#L25)
*   [code-block -> v19.0.1](https://github.com/ckeditor/ckeditor5/blob/893b873002c838fb10efda35515ccfaf44c3e39a/packages/ckeditor5-code-block/package.json#L16)
*   [paste-from-office -> v19.0.2](https://github.com/ckeditor/ckeditor5/blob/893b873002c838fb10efda35515ccfaf44c3e39a/packages/ckeditor5-paste-from-office/package.json#L30)

After updating everything and building GHW, all seemed ok until I’ve opened it in the browser and got a breaking error:

>  CKEditorError: ckeditor-duplicated-modules: Some CKEditor 5 modules are duplicated. Read more: [https://ckeditor.com/docs/ckeditor5/latest/framework/guides/support/error-codes.html#error-ckeditor-duplicated-modules](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/support/error-codes.html#error-ckeditor-duplicated-modules)

After some consultation with the team, we figured out that deleting node\_modules and yarn.lock and reinstalling all dependencies should solve the problem. This is not an obvious solution though and I'm afraid that others out there may face the same problem.

---

To sum up, I think that, considering that all packages are released under a single CKEditor 5 version - v19.1.1 in this case - all inter-dependencies in the packages of this release should point to the same versions - the ones included in the release, in fact.

**Comment by Reinmar** [comment]
Hm.. first of all, I think that we should not end up with what you @fredck described – all pkgs should be pointing to the same 19.0.2 version of utils. If it's a dependency of a plugin, that plugin should be released together with cke5-utils. It was our first ever limited patch release and we didn't have our tooling tested and well thought for that.

In other words, in this case, all pkgs would be released because whenever we need to have a patch release, we also need to release builds and for that we need to release ckeditor5-utils too as `CKEDITOR_VERSION` that's embedded in a build is read from that package. And since probably all packages rely on ckeditor5-utils (e.g. through ckeditor5-core), then we should release all the packages.

However, regardless of that, I'm leaning towards versioning all packages with the same version number. Semver makes things so much harder to analyse and less stable that any benefits it brings are lost.

**Comment by Reinmar** [comment]
We discussed this topic with @mlewand and @pomek and want to go with the same version for all packages regardless of the size of the change. 

Let's not forget to update: https://ckeditor.com/docs/ckeditor5/latest/framework/guides/support/versioning-policy.html.

**Comment by ma2ciek** [comment]
Have you been discussing what should happen with CF and CS versions in case of a bug fix in one of CKEditor 5 packages?

**Comment by Reinmar** [comment]
No, we focused on CKE5. Do you see any problems with e.g. CKE5 releasing 20.0.1 and CF being versioned as 20.0.0 (and referring to `^20.0.0` CKE5 packages)?

One issue that I can see is that people will use some CKE5 deps in 20.0.1 and some in 20.0.0 (collab packages). So, to avoid that we could have an empty release of CF packages. I'm not sure if this is worth the effort, though.
